### PR TITLE
feat(template): introduce rx-virtual-view

### PR DESCRIPTION
**Pull request by hoebbelsB**
# Description

A large number of DOM elements can significantly impact performance, leading to slow initial load times and sluggish interactions.

Especially mobile users have a very limited viewport available. Most of the pages contents are hidden below the fold. So why render them at all?

When dealing with large lists or data sets there is a technique, known as virtual scrolling or windowing. It drastically improves the performance of your Angular applications.

However, if you are not working with plain lists, or highly dynamic components, the concept of virtual scrolling isn't applicable. This is true for:

* masonry layouts
* dynamic grids
* landing pages with widgets

This is where the RxVirtualView directive comes in. It provides a simple way to only display the elements that are currently visible to the user.

image

## Usage

```html
<!-- use the root node (html) for the IntersectionObserver -->
<div rxVirtualViewObserver [root]="null">
  <div class="item" rxVirtualView>
    <!-- this will be the template when .item is visible to the user -->
    <div *rxVirtualViewTemplate>Virtual View 1</div>
    <!-- this will be the template when .item isn't visible to the user -->
    <div *rxVirtualViewPlaceholder>
      Placeholder
      <div></div>
    </div>
  </div>
</div>
```

```html

<div rxVirtualViewObserver class="container">
  @for (item of items; track item.id) {
  <div class="item" rxVirtualView>
    <item *rxVirtualViewTemplate [item]="item" />
    <div *rxVirtualViewPlaceholder style="height: 50px;"></div>
  </div>
  }
</div>

```

# Todos

* [x] docs
   * [x] feature intro
   * [x] minimal getting started
   * [ ] edge cases
      * [ ] startWithPlaceholderAsap
      * [ ] keepLastKnownSize
      * [ ] useContentVisibility
      * [ ] useContainment
* [x] jsDocs
* [x] error handling for missing providers (e.g. missing `VirtualViewObserver`)
* [x] rxjs 6.x compat
* [x] tests
* [x] api: trigger to "hide" all elements
  * e.g. you open up a modal on top of the whole page
* [x] configurable strategy for `showTemplate` & `showPlaceholder` respectively
* [x] flag as **developer preview**
* [x] default configuration via injection token for `RxVirtualView`
* [x] configurable cache 
* [ ] optional: reuse `RxaResizeObserver` from `experimental virtual-scrolling` -> extract to cdk

### CODE REVIEW

File: libs/template/virtual-view/src/lib/virtual-view-observer.directive.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    // Start observing the virtual view immediately.
+    this.#observer.observe(virtualView);
+
+    return combineLatest([isVisible$, this.#forcedHidden$]).pipe(
+      map(([isVisible, forcedHidden]) => (forcedHidden ? false : isVisible)),
+      startWith(false),
+      distinctUntilChanged(),
+      tap({
+        unsubscribe: () => this.#elements.delete(virtualView),
+      }),
```
Comment (Karnaukhov-kh) [inline libs/template/virtual-view/src/lib/virtual-view-observer.directive.ts:151]: rxjs 6 compat

File: libs/template/virtual-view/src/lib/virtual-view.directive.ts [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+   */
+  readonly cacheEnabled = input(this.#config.cacheEnabled, {
+    transform: booleanAttribute,
+  });
+
+  /**
+   * Whether to start with the placeholder asap or not.
+   *
+   * If `true`, the placeholder will be rendered immediately, without waiting for the template to be visible.
+   * This is useful when you want to render the placeholder immediately, but you don't want to wait for the template to be visible.
```
Comment (Karnaukhov-kh) [inline libs/template/virtual-view/src/lib/virtual-view.directive.ts:97]: This line is exactly what I would like to see in the main docs bullet points

File: libs/template/virtual-view/src/lib/virtual-view.directive.ts [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+                    this.size.set({
+                      width: borderBoxSize[0].inlineSize,
+                      height: borderBoxSize[0].blockSize,
+                    });
+                  }),
+                );
+          }
+          return this.#placeholderVisible() ? NEVER : this.showPlaceholder$();
+        }),
+        tap({
```
Comment (Karnaukhov-kh) [inline libs/template/virtual-view/src/lib/virtual-view.directive.ts:243]: rxjs 6 compat?

File: libs/template/virtual-view/src/lib/resize-observer.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import { Observable, ReplaySubject, Subject } from 'rxjs';
+import { distinctUntilChanged, tap } from 'rxjs/operators';
+
+/**
+ * A service that observes the resize of the elements.
+ *
+ * @developerPreview
+ */
+@Injectable()
+export class RxaResizeObserver {
```
Comment (Karnaukhov-kh) [inline libs/template/virtual-view/src/lib/resize-observer.ts:11]: Maybe also should go to cdk

File: apps/docs/docs/template/api/virtual-view-directive.mdx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+### Features
+
+- Performance Optimization: Only renders visible elements, improving initial load time and overall performance.
+- Template and Placeholder: Allows for defining a template for visible items and a placeholder for out-of-view items.
+- Observability: Observes the virtual view to determine which items are visible.
+- Configurability: Provides options for customizing behavior, such as rendering strategies and caching.
+
+### Usage
+
+The RxVirtualView directive is designed to work seamlessly with other related directives:
```
[the remark this answers is not in the document]
Comment (hoebbelsB) [reply apps/docs/docs/template/api/virtual-view-directive.mdx]: ty, updated

File: apps/docs/docs/template/api/virtual-view-directive.mdx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+This setup will:
+
+1. Use rxVirtualViewObserver to monitor the visibility of the rxVirtualView element.
+2. Render the content of rxVirtualViewTemplate when the element is visible.
+3. Show the rxVirtualViewPlaceholder when the element is not visible.
+
+### Advanced Usage
+
+You can further customize the behavior of RxVirtualView using these optional inputs:
```
[the remark this answers is not in the document]
Comment (hoebbelsB) [reply apps/docs/docs/template/api/virtual-view-directive.mdx:72]: yeah ...

File: libs/template/virtual-view/src/lib/virtual-view-cache.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import { VIRTUAL_VIEW_CONFIG_TOKEN } from './virtual-view.config';
+
+/**
+ * A service that caches templates and placeholders to optimize view rendering.
+ * It makes sure that all cached resources are cleared when the service is destroyed.
+ *
+ * @developerPreview
+ */
+@Injectable()
+export class VirtualViewCache implements OnDestroy {
```
[the remark this answers is not in the document]
Comment (hoebbelsB) [reply libs/template/virtual-view/src/lib/virtual-view-cache.ts:11]: I think at this point it's a rather specialized implementation for the RxVirtualView

File: libs/template/virtual-view/src/lib/resize-observer.ts (hunk 4 above)
Comment (hoebbelsB) [reply libs/template/virtual-view/src/lib/resize-observer.ts:11]: this one should def. go to CDK as it's also used in the virtual-scroll package

**Pull request merged**